shark_turbine/kernel/wave/codegen.py

```diff
@@ -541,6 +544,64 @@ def handle_mma(emitter: WaveEmitter, node: fx.Node):
     emitter.bind_node_proxy(node, IRProxyValue(result))
 
 
+@handle_op(shuffle)
+def handle_shuffle(emitter: WaveEmitter, node: fx.Node):
+    """
+    Generate gpu shuffle instruction to enable communication
+    between threads in a warp. Currently we only support
+    float unit vector that is <= 32 bits.
+
+    Translation to shuffle is done in 3 steps:
+    1. Scalarize (vector<1xf16> -> f16)
+    2. Pad to 32-bit if needed(f16 -> f32)
+    3. Shuffle (gpu.shuffle xor src, offset, width -> f32)
+    4. Reconstruct to original vector type (truncf f32 -> f16, broadcast -> vector<1xf16>)
+
+    TODO: Handle non-unit vector types such as vector<4xF8> (useful for resolving layouts).
+    """
+    try:
+        src, offset, width = node.args
+    except ValueError as e:
+        raise ValidationError("Malformed arguments") from e
+    if not isinstance(offset, int) or not isinstance(width, int):
+        raise NotImplementedError(
+            "Non-const wdith or offset is not yet implemented for shuffleOp."
+        )
+    src = cast_py_value(emitter, src).ir_value
+    offset = cast_py_value(emitter, offset, IntegerType.get_signless(32)).ir_value
+    width = cast_py_value(emitter, width, IntegerType.get_signless(32)).ir_value
+
+    if not VectorType.isinstance(src.type):
+        raise NotImplementedError("Scalar src is not implemented yet for shuffleOp.")
+
+    if math.prod(src.type.shape) != 1:
+        raise NotImplementedError("Currently only support unit vector for shuffleOp.")
+
+    # Scalarize (vector<FLOAT_TYPE> -> FLOAT_TYPE).
+    static_pos = [0 for i in range(src.type.rank)]
+    element = vector_d.extract(src, static_position=static_pos, dynamic_position=[])
+    element_original_type = element.type
+
+    # Pad to 32 bit if needed.
+    if not _is_float_type(element.type):
```

Comment:
If you look at the iree code, everything is always cast to i32 so you dont have to worry about support ints and floats. They both go down the same path.

Reply:
Yeap! In IREE we need to cast it to i32 for shuffle because Vulkan-SPIRV do not handle F32 shuffles. Although for iree-turbine, I'd prefer to have a separate path for int and float since it is relatively simple(just another if condition with exti) and it saves 2 bitcastings per shuffle! :)